File: 算法/代码随想录算法PDF/00调试源代码/763.划分字母区间.cpp
/*
 * @lc app=leetcode.cn id=763 lang=cpp
 *
 * [763] 划分字母区间
 */

// @lc code=start
class Solution {
public:
    static bool cmp(const vector<int>& a, const vector<int>& b) {
        return a[0] < b[0];
    }
    vector<vector<int>> getBord(string& s) {
        vector<vector<int>> hash(26, vector<int>(2, INT_MIN));
        for (int i = 0; i < s.size(); i++) {
            if (hash[s[i] - 'a'][0] == INT_MIN) hash[s[i] - 'a'][0] = i;  // 左边界
            hash[s[i] - 'a'][1] = i;  // 右边界
        }
        vector<vector<int>> hash_filter;  // 去掉没出现的字符
        for (int i = 0; i < 26; i++) {
            if (hash[i][0] != INT_MIN) hash_filter.push_back(hash[i]);
        }
        return hash_filter;
    }
    vector<int> partitionLabels(string s) {
        vector<vector<int>> hash = getBord(s);
        sort(hash.begin(), hash.end(), cmp);  // 按左区间从小到大排
        vector<int> res;
        int left = 0;
        int right = hash[0][1];
        for (int i = 1; i < hash.size(); i++) {
            if (hash[i][0] > right) {
                res.push_back(right - left + 1);
                left = hash[i][0];
            }
            right = max(right, hash[i][1]);
        }
        res.push_back(right - left + 1);  // 最后一个区间
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/763.划分字母区间.py
#
# @lc app=leetcode.cn id=763 lang=python3
#
# [763] 划分字母区间
#

# @lc code=start
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        # 记录每个字母出现的区间
        def getBord(s):
            hash = [[-float('inf')] * 2 for _ in range(26)]
            for i in range(len(s)):
                if hash[ord(s[i]) - ord('a')][0] == -float('inf'):
                    hash[ord(s[i]) - ord('a')][0] = i
                hash[ord(s[i]) - ord('a')][1] = i
            # 去除字符串中未出现的字母所占用区间
            hash_filter = []
            for item in hash:
                if item[0] != -float('inf'): hash_filter.append(item)
            return hash_filter

        # 得到无重叠区间题意中的输入样例格式：区间列表
        hash = getBord(s)
        # 按照左边界从小到大排序
        hash.sort(key= lambda x: x[0])
        res = []
        left = 0
        # 记录最大右边界
        right = hash[0][1]
        for i in range(len(hash)):
            # 一旦下一区间左边界大于当前右边界，即可认为出现分割点
            if hash[i][0] > right:
                res.append(right - left + 1)
                left = hash[i][0]
            # 实时更新最大右边界
            right = max(right, hash[i][1])
        # 最右侧区间（字符串长度为1时的特殊情况也包含于其中）
        res.append(right - left + 1)
        return res
# @lc code=end

